common/src/main/java/org/conscrypt/OpenSSLMac.java

```diff
@@ -137,57 +122,135 @@ protected void engineUpdate(ByteBuffer input) {
             throw new RuntimeException("Negative remaining amount");
         }
 
-        final NativeRef.HMAC_CTX ctxLocal = ctx;
-        NativeCrypto.HMAC_UpdateDirect(ctxLocal, ptr, len);
+        updateDirect(ptr, len);
         input.position(position + len);
     }
 
     @Override
     protected byte[] engineDoFinal() {
-        final NativeRef.HMAC_CTX ctxLocal = ctx;
-        final byte[] output = NativeCrypto.HMAC_Final(ctxLocal);
+        final byte[] output = doFinal();
         resetContext();
         return output;
     }
 
+    protected abstract byte[] doFinal();
+
     @Override
     protected void engineReset() {
         resetContext();
     }
 
-    public static final class HmacMD5 extends OpenSSLMac {
+    public static class Hmac extends OpenSSLMac {
+        private NativeRef.HMAC_CTX ctx;
+
+        /**
+         * Holds the EVP_MD for the hashing algorithm, e.g.
+         * EVP_get_digestbyname("sha1");
+         */
+        private final long evp_md;
+
+        public Hmac(long evp_md, int size) {
+            super(size);
+            this.evp_md = evp_md;
+        }
+
+        @Override
+        protected void resetContext() {
```

Comment:
Probably not for this CL since you need to refactor your init/reset logic, but both `HMAC_CTX` and `CMAC_CTX` support a more efficient reset for the same key. It avoids recomputing some partial state like AES key schedules or the first few HMAC blocks.

For CMAC, you call `CMAC_Reset`. For HMAC, you call, confusingly, not `HMAC_CTX_reset` but `HMAC_Init_ex(ctx, NULL, 0, NULL, NULL)`.

In particular, I noticed you all have to call `resetContext()` after `engineDoFinal()`, even if the object won't get called yet. The cheaper reset should avoid having to compute an extra key schedule and then throw it away.

As a bonus, this also means `keyBytes` no longer needs to be protected or retained. It can instead just be a parameter to some `initContext` method.

Reply:
Nice, will definitely do that in a future CL (and raise a bug so we remember to)